test_util/cluster.py

```diff
@@ -0,0 +1,260 @@
+        # Write out the ssh key to the local filesystem for the ssh lib to pick up.
+        with open(ssh_key_path, 'w') as ssh_key_fh:
+            ssh_key_fh.write(vpc.get_ssh_key())
+
+        cluster = cls(
+            vpc.hosts(),
+            ssh_user='centos',
+            ssh_key_path=ssh_key_path,
+            remote_dir='/home/centos',
+            num_masters=num_masters,
+            num_agents=num_agents,
+            num_public_agents=num_public_agents,
+            setup=setup,
+        )
+        cluster.vpc = vpc
+        return cluster
+
+    def cleanup(self):
+        if self.vpc is not None:
+            self.vpc.delete()
+
+    @staticmethod
+    def partition_cluster(hosts, num_masters, num_agents, num_public_agents):
+        """Return (bootstrap, masters, agents, public_agents) from hosts."""
+        assert num_masters in (1, 3, 5)
+        assert all(x >= 0 for x in [num_agents, num_public_agents])
+        assert num_agents + num_public_agents > 0
+        assert len(hosts) == 1 + num_masters + num_agents + num_public_agents
+
+        hosts_iter = iter(hosts)
+        return (
+            next(hosts_iter),
+            list(itertools.islice(hosts_iter, num_masters)),
+            list(itertools.islice(hosts_iter, num_agents)),
+            list(itertools.islice(hosts_iter, num_public_agents)),
+        )
+
+    def get_private_address(self, tunnel):
+        """Uses checked-in IP detect script to report private IP mapping
+        Args:
+            tunnel (SSHTunnel): see ssh.ssh_tunnel.SSHTunnel
+
+        Returns:
+            dict[public_IP] = private_IP
+        """
+        ip_detect_script = pkg_resources.resource_filename('gen', 'ip-detect/aws.sh')
+        tunnel.write_to_remote(ip_detect_script, join(self.remote_dir, 'ip-detect.sh'))
+        private_ip = tunnel.remote_cmd(['bash', join(self.remote_dir, 'ip-detect.sh')]).decode('utf-8').strip('\n')
+        assert len(private_ip.split('.')) == 4
+        return private_ip
+
+    def get_cluster_hosts(self, public_ips, setup):
+        ssh_hostports = [i + ':22' for i in public_ips]
+
+        @retry(stop_max_delay=120000)
+        def establish_host_connectivity():
+            """Continually try to recreate the SSH Tunnels to all hosts for 2 minutes
+            """
+            return closing(TunnelCollection(self.ssh_user, self.ssh_key_path, ssh_hostports))
+
+        logging.info('Checking that hosts are accessible')
+        with establish_host_connectivity() as tunnels:
+            private_ip = {}
+            for tunnel in tunnels.tunnels:
+                private_ip[tunnel.host] = self.get_private_address(tunnel)
+                if setup:
+                    # Make the default user priveleged to use docker
+                    tunnel.remote_cmd(['sudo', 'usermod', '-aG', 'docker', self.ssh_user])
+
+        return [Host(public_ip, private_ip[public_ip]) for public_ip in public_ips]
+
+    @contextmanager
+    def host_tunnel(self, host):
```

Comment:
why not just update SSHTunnel to have __enter__  + __exit__ since this is the only way we ever use it?